### PR TITLE
Make perf addon and instrumentation appear in production build

### PR DESCRIPTION
**Pull request by dmytromolkov**
Went around the codebase and found all if (__DEV__) statements that were hiding performance instrumentation from the production build.
I was trying to make sure that the other __DEV__ functionality (like additional validations and warnings) are still gated.

Tested it by building a new version and running it in my VM.

### CODE REVIEW

File: src/addons/transitions/ReactTransitionChildMapping.js [hunk 1]
```diff
@@ -27,11 +27,7 @@ var ReactTransitionChildMapping = {
       return children;
     }
 
-    if (__DEV__) {
-      return flattenChildren(children, selfDebugID);
-    }
-
-    return flattenChildren(children);
+    return flattenChildren(children, selfDebugID);
```
Comment (dgoldstein0) [inline src/addons/transitions/ReactTransitionChildMapping.js]: hm, why does this matter for perf instrumentation?

File: src/renderers/shared/stack/reconciler/ReactCompositeComponent.js [hunk 2]
```diff
@@ -370,25 +366,16 @@ var ReactCompositeComponent = {
...
+    ReactCurrentOwner.current = this;
+    try {
       return this._constructComponentWithoutOwner(
         doConstruct,
         publicProps,
         publicContext,
         updateQueue
       );
+    } finally {
+      ReactCurrentOwner.current = null;
```
Comment (dgoldstein0) [inline src/renderers/shared/stack/reconciler/ReactCompositeComponent.js]: why does ReactCurrentOwner matter?

File: src/renderers/shared/stack/reconciler/ReactMultiChild.js [hunk 3]
```diff
@@ -182,22 +180,17 @@ var ReactMultiChild = {
...
-        if (this._currentElement) {
-          try {
-            ReactCurrentOwner.current = this._currentElement._owner;
-            return ReactChildReconciler.instantiateChildren(
-              nestedChildren, transaction, context, selfDebugID
-            );
-          } finally {
-            ReactCurrentOwner.current = null;
-          }
+      var selfDebugID = getDebugID(this);
```
Comment (dgoldstein0) [inline src/renderers/shared/stack/reconciler/ReactMultiChild.js]: wondering why this is needed as well

File: src/renderers/shared/stack/reconciler/ReactMultiChild.js [hunk 4]
```diff
@@ -210,42 +203,28 @@ var ReactMultiChild = {
...
-          );
-          return nextChildren;
+
+      selfDebugID = getDebugID(this);
+      if (this._currentElement) {
+        try {
+          ReactCurrentOwner.current = this._currentElement._owner;
+          nextChildren = flattenChildren(nextNestedChildrenElements, selfDebugID);
+        } finally {
+          ReactCurrentOwner.current = null;
```
Comment (dgoldstein0) [inline src/renderers/shared/stack/reconciler/ReactMultiChild.js]: this chunk too

File: src/umd/ReactWithAddonsUMDEntry.js [hunk 5]
```diff
@@ -21,15 +21,13 @@ var ReactWithAddonsUMDEntry = Object.assign(ReactWithAddons, {
...
-      getNextDebugID: require('getNextDebugID'),
-    }
-  );
-}
+Object.assign(
+  ReactWithAddonsUMDEntry.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED,
+  {
+    // ReactComponentTreeHook should not be included in production.
+    ReactComponentTreeHook: require('ReactComponentTreeHook'),
+    getNextDebugID: require('getNextDebugID'),
```
Comment (dgoldstein0) [inline src/umd/ReactWithAddonsUMDEntry.js]: is it necessary to expose these?

File: src/umd/ReactUMDEntry.js [hunk 6]
```diff
@@ -20,15 +20,13 @@ var ReactUMDEntry = Object.assign(React, {
...
-      getNextDebugID: require('getNextDebugID'),
-    }
-  );
-}
+Object.assign(
+  ReactUMDEntry.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED,
+  {
+    // ReactComponentTreeHook should not be included in production.
+    ReactComponentTreeHook: require('ReactComponentTreeHook'),
+    getNextDebugID: require('getNextDebugID'),
```
Comment (dgoldstein0) [inline src/umd/ReactUMDEntry.js]: is it necessary to expose these?

**Review by dgoldstein0 — COMMENTED** [review]
mostly wondering why some bits are needed.  Not looking for things that you missed.

File: src/addons/transitions/ReactTransitionChildMapping.js (hunk 1 above)
Comment (dmytromolkov) [reply src/addons/transitions/ReactTransitionChildMapping.js]: I will trace it down further to make sure it is, but my understanding was that this debug ID is used by the instrumentation. So I was trying to make sure we are passing it along wherever we can.

**Comment by dmytromolkov** [comment]
Yeah, your comments were spot on. Gated those behind __DEV__ again

File: src/renderers/dom/client/ReactMount.js [hunk 7]
```diff
@@ -271,9 +267,8 @@ var TopLevelWrapper = function() {
   this.rootID = topLevelRootCounter++;
 };
 TopLevelWrapper.prototype.isReactComponent = {};
-if (__DEV__) {
-  TopLevelWrapper.displayName = 'TopLevelWrapper';
-}
+
+TopLevelWrapper.displayName = 'TopLevelWrapper';
```
Comment (dgoldstein0) [inline src/renderers/dom/client/ReactMount.js:271]: do we need this?

**Review by dgoldstein0 — APPROVED** [review]
overall this looks reasonable.  I don't have a ton of context though so I don't know if you missed anything or not.

**Comment by dgoldstein0** [comment]
also worth mentioning, it's unclear to me if we pulled in all the debugId code we need or not.  It does look like some of the instrumentation calls depend on it fwiw.

**Pull request merged**